kernel/examples/table-reader/src/main/java/io/delta/kernel/examples/SingleThreadedTableReader.java

```diff
@@ -94,18 +100,35 @@ private int readData(StructType readSchema, Scan scan, int maxRowCount) throws I
 
         int readRecordCount = 0;
         try {
+            StructType physicalReadSchema =
+                ScanStateRow.getPhysicalDataReadSchema(tableClient, scanState);
             while (scanFileIter.hasNext()) {
-                try (CloseableIterator<FilteredColumnarBatch> data =
-                    Scan.readData(
-                        tableClient,
-                        scanState,
-                        scanFileIter.next().getRows(),
-                        Optional.empty())) {
-                    while (data.hasNext()) {
-                        FilteredColumnarBatch dataReadResult = data.next();
-                        readRecordCount += printData(dataReadResult, maxRowCount - readRecordCount);
-                        if (readRecordCount >= maxRowCount) {
-                            return readRecordCount;
+                FilteredColumnarBatch scanFilesBatch = scanFileIter.next();
+                try (CloseableIterator<Row> scanFileRows = scanFilesBatch.getRows()) {
+                    while (scanFileRows.hasNext()) {
+                        Row scanFileRow = scanFileRows.next();
+                        FileStatus fileStatus =
+                            InternalScanFileUtils.getAddFileStatus(scanFileRow);
+                        CloseableIterator<ColumnarBatch> physicalDataIter =
+                            tableClient.getParquetHandler().readParquetFiles(
+                                singletonCloseableIterator(fileStatus),
+                                physicalReadSchema,
+                                Optional.empty());
+                        try (
+                            CloseableIterator<FilteredColumnarBatch> transformedData =
+                                Scan.transformPhysicalData(
```

Comment:
`Scan.transformPhysicalData` is a static method. Does that mean the connector has to re-parse, re-validate, and re-instantiate the transformation expression for every single batch? Or is there an engine expression stashed somewhere, that can be applied to every batch as it flies past?

Reply:
`Scan.transformPhysicalData` takes an iterator and returns a new iterator. Transformations are applied as the batches pass through, so that means we try to keep the state within the iterator (for example loading DV, we maintain the last loaded DV and see if it is still applicable for the next batch. If yes use it, otherwise load a new one).